docs(wallet): reword the `next_unused_address` doc

oleonardolima opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

Adds an example on what `used` stands for, and make it explicit that it
has the same behavior as `Wallet::reveal_next_address` in the scenario
where all previously revealed addresses have been used.

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

### Notes to the reviewers

Is there any other behavior of `next_unused_address` we'd need to make clear through documentation ? 

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

- Improve the `Wallet::next_unused_address` documentation to better describe expected behavior/usage.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [ ] I've added tests for the new feature
* [ ] I've added docs for the new feature

#### Bugfixes:

* [ ] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [ ] I'm linking the issue being fixed by this PR

notmandatory reviewed (dismissed):
ACK 07b85cade0af2d77a436252fbfabe3a724993b43

New name looks good to me.

evanlinjin commented:
Okay thinking about it, this method actually returns the earliest unused...

notmandatory commented:
lets make this a doc only change

oleonardolima commented:
As per last call discussion, I'll repurpose this PR to keep `next_unused_address` API, focusing only on improving the current documentation.

Ademan commented:
FWIW the docs tripped me up a bit, such that I was confused what the actual difference between `next_unused_address()` and `reveal_next_address()` was for a short while.

On subsequent readings it's **perfectly** clear, so I might have just been careless in my first reading...

But I _think_ the distinction between "used" and "derived" was not in my mind on my first reading.

```
    /// This will attempt to derive and reveal a new address if no newly revealed addresses
    /// are available. See also [`reveal_next_address`](Self::reveal_next_address).
```

Maybe instead of "available" it could be "unused" ? ("available" sent my mind to the generated key pools of pre-HD wallet bitcoin core)

and/or at the risk of repeating nearly the same thing as the first and second lines, add something _like_

"If any previously derived addresses are unused, `next_unused_address()` returns the one with the lowest derivation index, otherwise it derives a new address."

ValuedMammal commented:
@Ademan That's a great suggestion, thanks for weighing in on this. Here's another way we could word it as well:

```rust
    /// This will attempt to reveal a new address if all previously revealed addresses have
    /// been used, in which case the returned address will be the same as calling
    /// [`reveal_next_address`].
```

Ademan commented:
> @Ademan That's a great suggestion, thanks for weighing in on this. Here's another way we could word it as well:
> 
> ```rust
>     /// This will attempt to reveal a new address if all previously revealed addresses have
>     /// been used, in which case the returned address will be the same as calling
>     /// [`reveal_next_address`].
> ```

I like that! It directly contrasts the two calls.

As a side note, is there are reason why BDK uses the word "reveal" instead of "derive" in the API?

ValuedMammal commented:
> As a side note, is there are reason why BDK uses the word "reveal" instead of "derive" in the API?

When you create a wallet we first derive a number of addresses (script pubkeys) and store them internally. When you reveal an address, it comes from the pool of derived scripts. The concept of "revealing" exists for the purpose of keeping track of the highest index that we've uncovered. An address is considered "used" when you receive a transaction paying to that address.

oleonardolima committed:
docs(wallet): reword the `next_unused_address` doc

docs(wallet): reword the `next_unused_address` doc

Adds an example on what `used` stands for, and make it explicit that it
has the same behavior as `Wallet::reveal_next_address` in the scenario
where all previously revealed addresses have been used.

docs(wallet): fix typo in doc comment

oleonardolima commented:
Thanks for the suggestions.

I repurposed the PR to have only the documentation change: i) added a statement on what `used` stands for, and ii) applied the VM's suggestion in the second section.

thunderbiscuit reviewed (approved):
ACK 7ece14aca60f50c6ca1b4f083b9e72305f02da53.

oleonardolima commented:
@notmandatory @ValuedMammal do we want to get this one in for 1.0? Any other required changes?

ValuedMammal commented on crates/wallet/src/wallet/mod.rs:
```diff
@@ -736,10 +736,10 @@ impl Wallet {
 
     /// Get the next unused address for the given `keychain`, i.e. the address with the lowest
-    /// derivation index that hasn't been used.
+    /// derivation index that hasn't been used (hasn't been involved in any transaction).
```
```suggestion
    /// derivation index that hasn't been used in a transaction.
```

oleonardolima replied:
fixed in bcc5329afa8eae6493b54355063896be0f73ed9c

ValuedMammal reviewed (approved):
self-ACK d051e738

notmandatory commented:
@oleonardolima please just squash the commits and then we can merge this one.

oleonardolima commented:
> @oleonardolima please just squash the commits and then we can merge this one.

Done [b39cf08](https://github.com/bitcoindevkit/bdk/pull/1680/commits/b39cf08c98d9a5b53bb0eacd53c71c341ba37f76), sorry for the typo.

notmandatory reviewed (approved):
ACK b39cf08c98d9a5b53bb0eacd53c71c341ba37f76

ValuedMammal reviewed (approved):
ACK b39cf08c98d9a5b53bb0eacd53c71c341ba37f76
